File: CoreSkills/Sort/InsertionSort.py
# Definition for a pair.
# class Pair:
#     def __init__(self, key: int, value: str):
#         self.key = key
#         self.value = value
class Solution:
    def insertionSort(self, pairs: List[Pair]) -> List[List[Pair]]:
        ans = []
   
        # edge case
        if (pairs == []):
            return []

        # pairs[:]: create a copy of pairs (by default, python passes by reference)
        ans.insert(0, pairs[:])

        for i in range(1, len(pairs)):
            hold = pairs[i]

            j = 1
            while i - j >= 0 and pairs[i-j].key > hold.key:
                pairs[i-j+1] = pairs[i-j]
                j+=1
            
            pairs[i-j+1] = hold

            ans.insert(i, pairs[:])

        return ans

File: CoreSkills/Sort/InsertionSort.cpp
using namespace std;
# include <vector>
# include <string>

// Definition for a Pair
class Pair {
public:
    int key;
    string value;

    Pair(int key, string value) : key(key), value(value) {}
};

class Solution {
    public:
        vector<vector<Pair>> insertionSort(vector<Pair>& pairs) {
            vector<vector<Pair>> res;
    
            for (int i = 0; i < pairs.size(); i++) {
                Pair hold = pairs[i];
    
                int j;
                for (j = i-1; j >= 0 && pairs[j].key > hold.key; j--) {
                    pairs[j+1] = pairs[j];
                }        
                pairs[j+1] = hold;
    
                res.push_back(pairs);
            }
    
            return res;
        }
};
    